### DISCUSSION TITLE
Config: integrations.github

### DISCUSSION THREAD

**Discussion by DannyBen** [retypeapp/retype#102] [Ideas]
I know this is probably too late to do now, and will break existing deployments, but I can't help but feel that the `edit` and probably the `cname` config options should have been tucked away inside `integrations.github`:

```json
{
  "integrations": {
    "googleAnalytics": {
      "id": "..."
    },
    "github": {
      "cname": "jobly.dannyb.co",
      "edit": {
        "repo": "https://github.com/DannyBen/jobly-book/",
        "base": "src",
        "branch": "master"
      }
    }
  }
}
```

Rationale:

1. To my knowledge, GitHub is the only one using the `CNAME` file approach. If this is not the case, then it should stay where it is.
2. The `edit` link is not a general "edit pattern" (e.g. `http://github.com/path/{{ page }}`) - retype is converting it to a GitHub specific URL (`<repo>/blob/<branch>/<page>`).

Another thought that comes to mind - and this I believe is what mdbook does (not saying its better or worse, just an alternative):

1. No cname definition in config - instead, mdbook simply copies any non-md file to the target - so I can put a CNAME file myself, or `.whatever-conf.yml` if I deploy elsewhere.
2. Convert the current `edit` config, from an object, to a string pattern that includes a `{{ page }}` variable. This will simplify both the config and will make it generic (remember, right now the config contains three parts, and retype just inserts `blob` somewhere there).

**Comment by geoffreymcgill** [comment]
A bunch of work was done in `v1.8.0` related to both the `edit` and `cname` configs. 

The `edit` config now better handles GitLab and GitHub repo URL's. There was an [issue](https://github.com/supremegamers/wiki/issues/7) handling a GitLab URL which has been cleaned up. Also pointing to either the `/blob/` or `/edit/` points should work better. 

So, `edit` is less GitHub focused now (it was always meant to be generic) and should support any _edit_ endpoint.

Retype `v1.8.0` will have a new `url` project config for `retype.json` (and `retype.yml`!). If the `url` is set, the value will be used to automatically generate the `CNAME` file. This mostly makes the `cname` config obsolete.

You will be able to disable the `CNAME` file generation by setting `cname: false`. Or, you can override by setting `cname: example.com`.

> To my knowledge, GitHub is the only one using the CNAME file approach. If this is not the case, then it should stay where it is.

I'm not sure if GItLab supports `CNAME`, but I suspect they do or may in the future. I will do some research.

**Comment by geoffreymcgill** [comment]
> mdbook simply copies any non-md file to the target

In general, I reeeeealy do not like this approach, mostly from a security perspective. It makes it waaaaaaaaaay to easy to accidentally leak a file (especially proprietary source code files and such) that you definitely do not want being exposed publicly.

A docs project can be, and typically is, mixed right in with the main project source code, usually within a root `/docs` folder. The private source code for the project is sitting right next to the docs content, and it's not unreasonable to think someone could make a configuration mistake with Retype and accidentally expose something they shouldn't.

Retype does employ this auto-copy technique, but I feel is much more sophisticated. There are four scenarios where a file located in the [`input`](https://retype.com/configuration/project/#input) directory will be auto-copied into the `output` directory:

1. By default, any file type within the following list:

- `*.gif`
- `*.heif`
- `*.jpeg`
- `*.jpg`
- `*.png`
- `*.svg`
- `*.webp`
- `*.ai`
- `*.bmp`
- `*.eps`
- `*.pdf`
- `*.tiff`
- `*.txt`

All the above are considered content files.

2. By default, any file place within the following folders:

- `**/static/**`
- `**/public/**`
- `**/assets/**`
- `**/resources/**`

We consider the above folders as public. Anything placed within will be copied to the generated to the website.

3. **NEW in v1.8.0**: Within your content (in a `.md`) file, you have linked to file. That _linked to_ file will now be automatically copied to the correct path in the `output`. For instance, if you link to `sample.zip`, the `sample.zip` file will be copied to the `output`.

Within your `.md` content, you have linked to the file. You clearly meant for that link to be public and any visitor can click on the link, so the endpoint of the link is now also public. Retype makes sure that endpoint works, unless of course you have explicitly [excluded](https://retype.com/configuration/project/#exclude).

4. You have explicitly allowed with the [`include`](https://retype.com/configuration/project/#include) config

All the file, folders, and any other logic for how Retype auto-copies static files to the `output` folder is (or will be) published to the [`include`](https://retype.com/configuration/project/#include) config docs.

Specific files, file types, or directories can also be explicitly excluded by configuring [`exclude`](https://retype.com/configuration/project/#exclude).

Safety never takes a holiday. 🦺 

Hope this helps.

**Reply by DannyBen** [reply]
> In general, I reeeeealy do not like this approach

Yup. I agree the downsides are obvious.

> 3. NEW in v1.8.0: Within your content (in a .md) file, you have linked to file. That linked to file will now be automatically copied

That is cool!

I was aware of most of the other things you mentioned, since the docs are so concise and pleasant. But now that you mentioned it, I have an `assets` folder (which I created before knowing it is handled in a special way), but now that I see the list includes `resources` I wonder if I should put my images in `resources` so there will be only one "public assets" folder.

Is this a recommended approach? I mean, if it wasn't, you would have removed `resources` from that whitelist right?

On the other hand, I am uneasy about mixing my files with retype's.

**Reply by geoffreymcgill** [reply]
> I wonder if I should put my images in resources so there will be only one "public assets" folder.
> 
> Is this a recommended approach? I mean, if it wasn't, you would have removed resources from that whitelist right?

I don't think we have any specific recommended approach to where static resource files should be placed.

The best approach for each project may differ. If you have relatively few images and small to medium number of pages, then storing all the images in one directory would be good. It's easy to keep organized. 

As the number of pages grow and/or the number of images (or static files) grow, then you might want to store _closer_ to its content page. For instance, if I had a `/blog/` folder, I would likely create a `/blog/static/` sub-folder and put all the blog post images in there. 

If you have a lot of blog posts, you might want to organize by year. for instance, `/blog/2021/static/`. 

As `**/static/**` is one of the magic folder names, anywhere there is a folder named `static`, its entire contents will be copied to the correct position in the `output`.

Whether you call your folder `static`, `assets`, `resources`, `images`, `photos`, or `whatever`, it's up to you. There are no hard rules. We just wanted to ensure Retype is flexible enough to handle whatever scenario you throw at it.

This week, the team will be discussing the default use of the `/resources/` folder for the Retype static files.

Hope this helps.
